File: python/enolib/elements/fieldset.py
from ..errors.validation import Validation
from . import fieldset_entry
from . import section
from .element_base import ElementBase
from .missing import missing_fieldset_entry


class Fieldset(ElementBase):
    def __init__(self, context, instruction, parent=None):
        super().__init__(context, instruction, parent)

        self._all_entries_required = parent._all_elements_required if parent else False

    def __repr__(self):
        return f"<class Fieldset key={self._instruction['key']} entries={len(self._entries())}>"

    def _entries(self, as_map=False):
        if not hasattr(self, '_instantiated_entries'):
            self._instantiated_entries = []
            self._instantiated_entries_map = {}
            self._instantiate_entries(self._instruction)

        return self._instantiated_entries_map if as_map else self._instantiated_entries

    def _entry(self, key, *, required=False):
        self._touched = True

        if not key:
            entries = self._entries()
        else:
            entries_map = self._entries(True)
            entries = entries_map[key] if key in entries_map else []

        if len(entries) == 0:
            if required or self._all_entries_required:
                raise Validation.missing_element(self._context, key, self._instruction, 'missing_fieldset_entry')
            elif required is None:
                return missing_fieldset_entry.MissingFieldsetEntry(key, self)
            else:
                return None

        if len(entries) > 1:
            raise Validation.unexpected_multiple_elements(
                self._context,
                key,
                [e._instruction for e in entries],
                'expected_single_fieldset_entry'
            )

        return entries[0]

    def _instantiate_entries(self, fieldset):
        if 'mirror' in fieldset:
            self._instantiate_entries(fieldset['mirror'])
        elif 'entries' in fieldset:
            def instantiate_and_index(entry):
                instance = fieldset_entry.FieldsetEntry(self._context, entry, self)

                if entry['key'] in self._instantiated_entries_map:
                    self._instantiated_entries_map[entry['key']].append(instance)
                else:
                    self._instantiated_entries_map[entry['key']] = [instance]

                return instance

            filtered = [entry for entry in fieldset['entries'] if entry['key'] not in self._instantiated_entries_map]
            native_entries = [instantiate_and_index(entry) for entry in filtered]

            if 'extend' in fieldset:
                self._instantiate_entries(fieldset['extend'])

            self._instantiated_entries.extend(native_entries)

    def _missingError(self, entry):
        raise Validation.missing_element(self._context, entry._key, self._instruction, 'missing_fieldset_entry')

    def _untouched(self):
        if not hasattr(self, '_touched'):
            return self._instruction

        untouched_entry = next((entry for entry in self._entries() if not hasattr(entry, '_touched')), None)

        return untouched_entry._instruction if untouched_entry else False

    def all_entries_required(self, required=True):
        self._all_entries_required = required

    def assert_all_touched(self, message=None, *, only=None, skip=None):
        entries_map = self._entries(True)

        for key, entries in entries_map.items():
            if (skip and key in skip) or (only and key not in only):
                continue

            for entry in entries:
                if not hasattr(entry, '_touched'):
                    if callable(message):
                        message = message(entry)

                    raise Validation.unexpected_element(self._context, message, entry['instruction'])

    def entries(self, key=None):
        self._touched = True

        if not key:
            return self._entries()
        else:
            entries_map = self._entries(True)
            return entries_map[key] if key in entries_map else []

    def entry(self, key=None):
        return self._entry(key)

    def optional_entry(self, key=None):
        return self._entry(key, required=False)

    def parent(self):
        return self._parent or section.Section(self._context, self._instruction['parent'])

    def required_entry(self, key=None):
        return self._entry(key, required=True)

    def touch(self):
        self._touched = True

        for entry in self._entries():
            entry._touched = True

File: ruby/lib/enolib/elements/fieldset.rb
# frozen_string_literal: true

module Enolib
  class Fieldset < ElementBase
    attr_reader :instruction, :touched

    def initialize(context, instruction, parent = nil)
      super(context, instruction, parent)

      @all_entries_required = parent ? parent.all_elements_required? : false
    end

    def _missing_error(entry)
      raise Errors::Validation.missing_element(@context, entry.key, @instruction, 'missing_fieldset_entry')
    end

    def _untouched
      return @instruction unless instance_variable_defined?(:@touched)

      untouched_entry = _entries.find { |entry| !entry.instance_variable_defined?(:@touched) }

      untouched_entry ? untouched_entry.instruction : false
    end

    def all_entries_required(required = true)
      @all_entries_required = required
    end

    def assert_all_touched(message = nil, except: nil, only: nil, &block)
      message = Proc.new(&block) if block_given?

      _entries(map: true).each do |key, entries|
        next if except && except.include?(key) || only && !only.include?(key)

        entries.each do |entry|
          next if entry.touched  # TODO: Revisit instance var existance question in ruby

          if message.is_a?(Proc)
            message = message.call(entry)
          end

          raise Errors::Validation.unexpected_element(@context, message, entry[:instruction])
        end
      end
    end

    def entries(key = nil)
      @touched = true

      if key
        entries_map = _entries(map: true)
        entries_map.has_key?(key) ? entries_map[key] : []
      else
        _entries
      end
    end

    def entry(key = nil)
      _entry(key)
    end

    def optional_entry(key)
      _entry(key, required: false)
    end

    def parent
      @parent || Section.new(@context, @instruction[:parent])
    end

    def required_entry(key = nil)
      _entry(key, required: true)
    end

    def to_s
      "#<Enolib::Fieldset key=#{@instruction[:key]} entries=#{_entries.length}>"
    end

    def touch
      @touched = true

      _entries.each(&:touch)
    end

    private

    def _entries(map: false)
      unless instance_variable_defined?(:@instantiated_entries)
        @instantiated_entries = []
        @instantiated_entries_map = {}
        instantiate_entries(@instruction)
      end

      map ? @instantiated_entries_map : @instantiated_entries
    end

    def _entry(key = nil, required: nil)
      @touched = true

      if key
        entries_map = _entries(map: true)
        entries = entries_map.has_key?(key) ? entries_map[key] : []
      else
        entries = _entries
      end

      if entries.empty?
        if required || required == nil && @all_entries_required
          raise Errors::Validation.missing_element(@context, key, @instruction, 'missing_fieldset_entry')
        elsif required == nil
          return MissingFieldsetEntry.new(key, self)
        else
          return nil
        end
      end

      if entries.length > 1
        raise Errors::Validation.unexpected_multiple_elements(
          @context,
          key,
          entries.map(&:instruction),
          'expected_single_fieldset_entry'
        )
      end

      entries[0]
    end

    def instantiate_entries(fieldset)
      if fieldset.has_key?(:entries)
        filtered = fieldset[:entries].reject { |entry| @instantiated_entries_map.has_key?(entry[:key]) }
        native_entries = filtered.map do |entry|
          instance = FieldsetEntry.new(@context, entry, self)

          if @instantiated_entries_map.has_key?(entry[:key])
            @instantiated_entries_map[entry[:key]].push(instance)
          else
            @instantiated_entries_map[entry[:key]] = [instance]
          end

          instance
        end

        @instantiated_entries.concat(native_entries)
      end
    end
  end
end
